### PR TITLE
fix: stabilize workspace search and conversation cache

### PR DESCRIPTION
**Pull request by guyi2000**
## Summary

- Problem: `grep_files` could abort when a broad search encountered unsupported/no-extension files, restored tool turns could show duplicate assistant messages, and large local conversation archives could exceed browser `localStorage` quota.
- Why it matters: these failures interrupt workspace-tool conversations and can make the chat UI unstable after larger grep/glob/tool-result sessions.
- What changed: workspace grep now skips unsupported/unreadable/oversized files; graph snapshot persistence now appends only tool messages after the summary pair; frontend archive saves compact oversized local cache on quota errors.
- What did NOT change (scope boundary): no new tool permissions, no schema changes, no changes to backend conversation/database source of truth.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [x] Gateway / orchestration
- [x] Skills / tool execution
- [ ] Auth / tokens
- [x] Memory / storage
- [ ] Integrations
- [x] API / contracts
- [x] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes N/A
- Related N/A
- [x] This PR fixes a bug or regression

## Root Cause / Regression History (if applicable)

- Root cause: broad workspace grep collected files before extension filtering, then threw on unsupported extensions; conversation snapshot persistence mixed summary persistence with raw LangGraph message deltas; frontend local archive writes assumed the serialized cache would fit in `localStorage`.
- Missing detection / guardrail: no regression coverage for unsupported grep files, graph snapshot persistence after stream summary persistence, or quota fallback for the local conversation archive.
- Prior context (`git blame`, prior PR, issue, or refactor if known): Unknown.
- Why this regressed now: richer tool-result persistence and local snapshots made large workspace-tool sessions more likely to hit these edges.
- If unknown, what was ruled out: the backend DB remains the source of truth; the quota failure is isolated to the browser cache key `structureclaw.console.conversations`.

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [x] Unit test
  - [x] Seam / integration test
  - [ ] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file: `backend/src/agent-langgraph/__tests__/workspace-tools.test.mjs`, `backend/tests/chat.route.test.mjs`, `frontend/tests/integration/console-page.test.tsx`.
- Scenario the test should lock in: unsupported/no-extension files do not fail grep; stream persistence does not duplicate final assistant messages; quota errors compact the local conversation cache.
- Why this is the smallest reliable guardrail: each test isolates the failing boundary without requiring an LLM or full browser E2E flow.
- Existing test that already covers this (if any): N/A.
- If no new test is added, why not: N/A.

## User-visible / Behavior Changes

- Workspace grep searches should no longer abruptly stop when broad patterns encounter unsupported files.
- Restored tool conversations should no longer show the same assistant turn twice.
- If the browser conversation cache grows too large, the UI now compacts local-only cached data instead of throwing `QuotaExceededError`. Backend conversation history and snapshots remain the primary persisted state.

## Diagram (if applicable)

```text
Before:
[large grep/tool session] -> [unsupported file or huge local archive] -> [tool abort / UI storage error]
[stream summary persisted] -> [full graph assistant persisted again] -> [duplicate restored assistant bubble]

After:
[large grep/tool session] -> [skip unsupported/cache-heavy payloads] -> [search/UI continues]
[stream summary persisted] -> [tool messages only appended] -> [single restored assistant turn]
```

## Security Impact (required)

- New permissions/capabilities? (`Yes/No`) No
- Secrets/tokens handling changed? (`Yes/No`) No
- New/changed network calls? (`Yes/No`) No
- Command/tool execution surface changed? (`Yes/No`) No
- Data access scope changed? (`Yes/No`) No
- If any `Yes`, explain risk + mitigation: N/A

### Environment

- OS: Windows / PowerShell
- Runtime/container: local Node.js workspace
- Model/provider: N/A
- Integration/channel (if any): Fastify backend + Next.js frontend tests
- Relevant config (redacted): default local development config

### Steps

1. Run `grep_files` with the default broad pattern in a workspace containing a no-extension file such as `LICENSE`.
2. Stream a tool-using conversation and restore it from backend persisted messages.
3. Save an oversized `structureclaw.console.conversations` local cache payload.

### Expected

- Grep skips unsupported files and returns supported matches.
- Restored chat shows one assistant response for the turn plus tool messages.
- Browser cache save compacts local-only data instead of throwing `QuotaExceededError`.

### Actual

- Before this PR, those scenarios could abort tool execution, duplicate an assistant turn, or throw a storage quota exception.

## Evidence

- [x] Failing test/log before + passing after
- [ ] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

Passing after:

```text
npm test --prefix backend -- --runInBand src/agent-langgraph/__tests__/workspace-tools.test.mjs tests/chat.route.test.mjs
PASS tests/chat.route.test.mjs
PASS src/agent-langgraph/__tests__/workspace-tools.test.mjs
Test Suites: 2 passed, 2 total
Tests: 9 passed, 9 total

npm run type-check --prefix frontend
passed

npm run lint --prefix frontend
passed with pre-existing <img> warnings in ai-console.tsx
```

## Human Verification (required)

- Verified scenarios: targeted backend workspace-tool and chat-route regression tests; frontend type-check; frontend lint.
- Edge cases checked: no-extension file during grep, existing tool-message count during snapshot persistence, oversized local cache save path.
- What you did **not** verify: full frontend integration test execution; the integration runner's test backend did not start within its 30 second fixture timeout in this local environment.

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [ ] I left unresolved only the conversations that still need reviewer or maintainer judgment.

If a bot review conversation is addressed by this PR, resolve that conversation yourself. Do not leave bot review conversation cleanup for maintainers.

## Compatibility / Migration

- Backward compatible? (`Yes/No`) Yes
- Config/env changes? (`Yes/No`) No
- Migration needed? (`Yes/No`) No
- If yes, exact upgrade steps: N/A

## Risks and Mitigations

- Risk: local browser archive compaction can remove local-only debug details and visualization snapshots after quota pressure.
  - Mitigation: backend conversation messages/snapshots remain the source of truth, and compaction only runs after a quota failure.

### CODE REVIEW

File: backend/src/agent-langgraph/workspace-tools.ts [hunk 1]
```diff
@@ -170,12 +176,32 @@ export function createGrepFilesTool() {
       const offset = Math.max(input.offset ?? 0, 0);
       const maxResults = Math.min(input.maxResults ?? 50, 100);
       const needle = input.query.toLowerCase();
+      let skippedFiles = 0;
       for (const rel of files.sort()) {
         const full = safeResolve(root, rel);
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-langgraph/workspace-tools.ts:183]: Addressed in `de0dfce`: `grep_files` now uses a shared `MAX_GREP_MATCHES` cap and breaks the outer file loop once the cap is reached, so it stops doing additional file I/O. I also added a regression test that places an unsupported file after a 1000-match file and asserts it is not scanned.

**Review by qinsz01 — APPROVED** [review]
